File: Nqueens.java
// Time Complexity : O(N^N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :

import java.util.*;
class Solution {
    List<List<String>> result;
    public List<List<String>> solveNQueens(int n) {
        result = new ArrayList<>();
        int[][] board = new int[n][n];
        recurse(n, 0, board);
        return result;
    }

    public void recurse(int n, int row, int[][] board){
        if(row == n){
            List<String> val = new ArrayList<>();
            for(int i=0; i<n; i++){
                StringBuilder temp = new StringBuilder();
                for(int j=0; j<n; j++){
                    if(board[i][j] == 1){
                        temp.append("Q");
                    }
                    else {
                        temp.append(".");
                    }
                }
                val.add(temp.toString());
            }
            this.result.add(val);
            return;
        }
        // System.out.println
        for(int i=0; i<n; i++){
            if(isSafe(board, new int[] {row, i})){
                board[row][i] = 1;
                recurse(n, row+1, board);
                board[row][i] = 0;
            }
        }

    }

    public boolean isSafe(int[][] board, int[] pos){
        int row = pos[0];
        int col = pos[1];
        for(int i= row; i>=0; i--){
            if(board[i][col] == 1)
                return false;
        }
        
        while(row >=0 && col >=0){
            if(board[row][col] == 1)
                return false;
            row--;
            col--;
        }
        row = pos[0];
        col = pos[1];
        while(row >=0 && col < board[0].length){
            if(board[row][col] == 1)
                return false;
            row--;
            col++;
        }
        return true;
    }
}

File: NQueens.cs
// Time Complexity : O(n!), n chices in first rown, n-1 in second row, n-2 in third row and so on
// Space Complexity : O(n^2), we need to mantain grid of n^2 size.
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes, debugged and found issue


// Your code here along with comments explaining your approach

List<IList<string>> result;
    bool[,] grid;
    public IList<IList<string>> SolveNQueens(int n) {
        
        result = new List<IList<string>>();
        if(n == 0)
            return result;
        
        grid = new bool[n,n];
        backtracking(0);
        
        return result;
    }
    
    private void backtracking(int row)
    {
        //base
        if(row == grid.GetLength(0))
        {
            List<string> newLst = new List<string>();
            for(int i = 0; i < grid.GetLength(0); i++)
            {
                StringBuilder sb = new StringBuilder();
                for(int j = 0; j < grid.GetLength(1); j++)
                {
                    if(grid[i,j])
                        sb.Append("Q");
                    else
                        sb.Append(".");
                }
                newLst.Add(sb.ToString());
            }
            result.Add(newLst);
            return;
        }
        
        //logic
        for(int i = 0; i < grid.GetLength(1); i++)
        {
            if(isSafe(row, i))
            {
                //action
                grid[row,i] = true;
                
                //recurse
                backtracking(row+1);
                
                //backtrack
                grid[row,i] = false;
            }
        }
    }
    
    private bool isSafe(int r, int c)
    {
        int i = 0;
         //check th eabove column to see if any other queen is present.
        for(i = 0; i < r; i++)
        {
            if(grid[i,c] == true)
            return false;
        }
        
        i = r;
        int j = c;
        //check th upper right
        while(i >= 0 && j < grid.GetLength(1))
        {
            if(grid[i, j] == true)
                return false;
            i--;
            j++;
        }

        i = r;
        j = c;
        //check the upper left
        while(i >= 0 && j >= 0)
        {
            if(grid[i,j] == true)
                return false;
            i--;
            j--;
        }
        return true;
    }